src/agent/coverage/src/code.rs

```diff
@@ -0,0 +1,331 @@
+                };
+
+                let inserted = symbols.index.insert(entry.clone());
+                if !inserted {
+                    log::error!("failed to insert symbol index entry: {:x?}", entry);
+                }
+            }
+        }
+
+        Ok(Self {
+            path,
+            base_va,
+            symbols,
+        })
+    }
+}
+
+#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
+#[serde(transparent)]
+pub struct SymbolIndex {
+    pub index: RegionIndex<Symbol>,
+}
+
+impl SymbolIndex {
+    pub fn iter(&self) -> impl Iterator<Item = &Symbol> {
+        self.index.iter()
+    }
+
+    /// Find the symbol metadata for the image-relative `offset`.
+    pub fn find(&self, offset: u64) -> Option<&Symbol> {
+        self.index.find(offset)
+    }
+}
+
+#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
+pub struct Symbol {
+    /// Raw symbol name, possibly mangled.
+    pub name: String,
+
+    /// File offset of the symbol definition in the on-disk module.
+    pub file_offset: u64,
+
+    /// Module-relative offset of the mapped symbol.
+    pub image_offset: u64,
+
+    /// Total size in bytes of the symbol definition.
+    pub size: u64,
+}
+
+impl Symbol {
+    pub fn contains_file_offset(&self, offset: u64) -> bool {
+        let lo = self.file_offset;
+        let hi = lo + self.size;
+        (lo..hi).contains(&offset)
+    }
+
+    pub fn contains_image_offset(&self, offset: u64) -> bool {
+        let lo = self.image_offset;
+        let hi = lo + self.size;
+        (lo..hi).contains(&offset)
+    }
+}
+
+/// Symbol metadata defines a `Region` relative to its process image.
+impl Region for Symbol {
+    fn base(&self) -> u64 {
+        self.image_offset
+    }
+
+    fn size(&self) -> u64 {
+        self.size
+    }
+}
+
+#[derive(Clone, Debug, Default, Deserialize, Serialize)]
+#[serde(transparent)]
+pub struct SymbolFilterSpec {
+    filters: HashMap<String, Filter>,
+}
+
+#[derive(Clone, Debug)]
+pub struct SymbolFilter {
+    /// Pre-compiled regex set for fast matching.
+    regexes: RegexSet,
+
+    /// Maps module name regexes to filters for symbol names.
+    filters: BTreeMap<usize, Filter>,
+}
+
+impl SymbolFilter {
+    pub fn new(mut spec: SymbolFilterSpec) -> Result<Self> {
+        let regexes = RegexSet::new(spec.filters.keys())?;
+
+        let mut filters = BTreeMap::default();
+
+        for (idx, pat) in regexes.patterns().iter().enumerate() {
+            // Guaranteed by construction.
+            let filter = spec.filters.remove(pat).unwrap();
+
+            filters.insert(idx, filter);
+        }
+
+        Ok(Self { regexes, filters })
+    }
+
+    pub fn is_allowed(&self, module: &ModulePath, name: &str) -> bool {
+        let rules = self.regexes.matches(&module.path_lossy());
+
+        // Check if there is some symbol rule for the module path.
+        //
+        // If many rules would apply, the first rule is used.
```

Comment:
Just wondering if there is any concern about impact of the ordering of the rules
It seems that this function could give a different result depending on which rule came first.

Reply:
It is definitely order-dependent. I consider it reasonable because the order is exposed to and fully controlled by the user.

My thinking was:
1. We're letting users provide multiple regexes to match symbols and modules (vs. _exact_ names).
2. We cannot statically know if the regexes will overlap.
3. By (1) and (2), we need to be able to handle conflicts / multiple matches.
4. We want the conflict handling to be predictable by the user, and debuggable.

Deferring to the user-provided "definition order" seemed like it'd be easy to understand and document. I'm open to other ideas!

Reply:
Actually, the outer _module_ matching is not fixed and user-controlled, it is controlled by the implementation of `HashMap`. I'm going to revisit this to make my above comments true at the module level, too.

Reply:
Will address in a follow-up PR.